contrib/binutils/patches/int.patch

```diff
@@ -0,0 +1,23 @@
+fixes edge case of `li t1, 9223372036854775807` in riscv64 assembly
+diff --git a/gas/config/tc-riscv.c b/gas/config/tc-riscv.c
+index a4161420128..0c31b2fa311 100644
+--- a/gas/config/tc-riscv.c
++++ b/gas/config/tc-riscv.c
+@@ -1998,7 +1998,7 @@ load_const (int reg, expressionS *ep)
+   bfd_vma upper_imm, sign = (bfd_vma) 1 << (RISCV_IMM_BITS - 1);
+   expressionS upper = *ep, lower = *ep;
+   lower.X_add_number = ((ep->X_add_number & (sign + sign - 1)) ^ sign) - sign;
+-  upper.X_add_number -= lower.X_add_number;
++  upper.X_add_number = (bfd_vma)upper.X_add_number - lower.X_add_number;
+ 
+   if (ep->X_op != O_constant)
+     {
+@@ -2009,7 +2009,7 @@ load_const (int reg, expressionS *ep)
+   if (xlen > 32 && !IS_SEXT_32BIT_NUM (ep->X_add_number))
+     {
+       /* Reduce to a signed 32-bit constant using SLLI and ADDI.  */
+-      while (((upper.X_add_number >> shift) & 1) == 0)
++      while ((((bfd_vma)upper.X_add_number >> shift) & 1) == 0)
```

Comment:
without this, the loop runs indefinitely until shift overflows

is this actually due to `int`? i'm not sure about that

Reply:
more likely clang optimises the ub so it does that

you should send it upstream

Reply:
ok, i'll send the changes upstream soon